### PR TITLE
Debugging Alternative New ShardMap Implementation

### PR DESCRIPTION
**Pull request by gerardtoconnor**
This is not a real PR, I was advised to do a PR to start a discussion on issues I'm having so people can provide insight. (Everything builds but some tests fail with no debug insight)

I have created a hybrid Map that is both immutable and hash-based such that lookups are a lot faster than Map, with the only situation where it is slower is folding (and although iteration is faster, it is not ordered). Fold & Foldback are also in same random order so I think this might be why it is breaking but in cases I have looked through, the order did not matter for fold operations (normally creating a map).

If anyone knows anywhere in complier where enumeration or folding maps matters for order, I believe will help find the issue as to why some tests failing (most pass).

The map is designed to be a hash space Map like Dictionary but the hash space is sliced up into shards so that mutation operations can swap shards in and out reducing the size of the base hash space being copied. The buckets are MapTree nodes to allow and continue immutability operations while minimising lookup time on hash conflict. The GC is performant enough that the overhead of creating shards (16x ref) is not hindering performance and size (memory usage) is only slightly bigger from crude tests. Also, alot of batch operations like Map.ofList etc mean that the hash base can be kept in the one atomic operation as well as not needing to swap out shards.

The segregated shards in their unique hash space allow for a lot of parallel operations that are difficult with normal Maps so for complex operations of Mapping & Union, it can be very quick.

I have included some perf tests below from [tests repo](https://github.com/gerardtoconnor/MapTests/blob/master/Tests.fs)

Immutable add new very slow for Dictionary as expected, fastest on ShardMap:

|          Method |        Mean |      Error |     StdDev | Scaled | ScaledSD |
|---------------- |------------:|-----------:|-----------:|-------:|---------:|
| ShardMap_AddNew |    495.3 ns |   5.744 ns |   5.091 ns |   1.00 |     0.00 |
|     Bmap_AddNew |    900.9 ns |   5.450 ns |   5.098 ns |   1.82 |     0.02 |
|     Dict_AddNew | 30,979.1 ns | 315.670 ns | 295.278 ns |  62.55 |     0.84 |

Creating Dictionary fastest from collection, BMap alot slower due to resizing/shifting:

|          Method |      Mean |      Error |     StdDev | Scaled | ScaledSD |    Gen 0 |  Gen 1 | Allocated |
|---------------- |----------:|-----------:|-----------:|-------:|---------:|---------:|-------:|----------:|
| Create_ShardMap |  86.67 us |  1.1059 us |  0.9804 us |   1.00 |     0.00 |  13.5498 |      - |  55.58 KB |
|     Create_BMap | 779.00 us | 14.8665 us | 13.9062 us |   8.99 |     0.18 | 125.0000 | 8.7891 | 512.98 KB |
|     Create_Dict |  32.24 us |  0.2768 us |  0.2454 us |   0.37 |     0.00 |   6.8359 |      - |  30.29 KB |

Shardmap marginly quicker as it breaks enumerator on success:

|             Method |      Mean |     Error |    StdDev |
|------------------- |----------:|----------:|----------:|
| ShardMap_ExistsPar | 13.988 us | 0.1215 us | 0.1077 us |
|    ShardMap_Exists |  9.724 us | 0.0823 us | 0.0770 us |
|        BMap_Exists | 15.144 us | 0.3100 us | 0.4139 us |

Union of multiple maps significantly faster on shardmap thanks to lookup speed:

|         Method |       Mean |     Error |    StdDev | Scaled | ScaledSD |    Gen 0 |    Gen 1 |   Gen 2 |  Allocated |
|--------------- |-----------:|----------:|----------:|-------:|---------:|---------:|---------:|--------:|-----------:|
| ShardMap_Union |   667.4 us |  13.13 us |  14.05 us |   1.00 |     0.00 | 131.0547 |  59.3099 |       - |  152.25 KB |
|     BMap_Union | 5,041.8 us | 123.20 us | 120.99 us |   7.56 |     0.23 | 492.1875 | 125.0000 | 54.6875 | 3113.27 KB |

Fold slower as Shardmap hops through empty buckets:

|              Method |     Mean |     Error |    StdDev | Scaled |
|-------------------- |---------:|----------:|----------:|-------:|
|       ShardMap_Fold | 13.88 us | 0.1117 us | 0.1045 us |   1.00 |
| ShardMap_FoldReduce | 11.65 us | 0.0628 us | 0.0587 us |   0.84 |
|           BMap_Fold | 11.76 us | 0.0952 us | 0.0844 us |   0.85 |

ShardMap alot faster as it can find slot quickly with hashing from list of repeating keys:

|             Method |      Mean |     Error |    StdDev | Scaled | ScaledSD |
|------------------- |----------:|----------:|----------:|-------:|---------:|
| ShardMap_LayerList |  33.79 us | 0.6663 us | 0.7673 us |   1.00 |     0.00 |
|     BMap_LayerList | 211.84 us | 0.9110 us | 0.8522 us |   6.27 |     0.14 |

checking all entries added to map fastest for Dictionary but not far off for ShardMap:

|                                      Method |      Mean |     Error |    StdDev | Scaled | ScaledSD |
|-------------------------------------------- |----------:|----------:|----------:|-------:|---------:|
|               ShardMap_contains_all_entries |  76.76 us | 0.7592 us | 0.7101 us |   1.00 |     0.00 |
|   ShardMap_contains_all_entries_TryGetValue |  66.10 us | 0.5044 us | 0.4471 us |   0.86 |     0.01 |
| ShardMap_contains_all_entries_FindOptStruct |  70.22 us | 0.8362 us | 0.7413 us |   0.91 |     0.01 |
|                   BMap_contains_all_entries | 197.51 us | 1.5434 us | 1.4437 us |   2.57 |     0.03 |
|                   Dict_contains_all_entries |  53.96 us | 0.4011 us | 0.3752 us |   0.70 |     0.01 |

Mapping a map to new map fastest with shardmap thanks to parrallelism:

|       Method |     Mean |    Error |   StdDev |   Median | Scaled | ScaledSD |
|------------- |---------:|---------:|---------:|---------:|-------:|---------:|
| ShardMap_Map | 104.3 us | 1.377 us | 1.075 us | 104.4 us |   1.00 |     0.00 |
|     Bmap_Map | 205.5 us | 4.123 us | 8.605 us | 202.2 us |   1.97 |     0.08 |
|     Dict_Map | 211.2 us | 4.273 us | 6.395 us | 208.8 us |   2.03 |     0.06 |

Sequence itteration a but faster then Bmap (but unordered) Dictionary way faster

|           Method |      Mean |     Error |    StdDev | Scaled | ScaledSD |
|----------------- |----------:|----------:|----------:|-------:|---------:|
| ShardMap_SeqIttr | 44.668 us | 0.8704 us | 1.3551 us |   1.00 |     0.00 |
|     BMap_SeqIttr | 55.762 us | 0.8101 us | 0.7578 us |   1.25 |     0.04 |
|     Dict_SeqIttr |  5.652 us | 0.0762 us | 0.0713 us |   0.13 |     0.00 |

@dsyme @vasily-kirichenko @forki

### CODE REVIEW

**Comment by gerardtoconnor** [comment]
I know some tidying and refactoring is required so apologies for some of commenting out of code and duplicated code but would naturally have this all tidied before an actual pull, just want to figure out issues first given all basic operations work fine outside of compiler.

**Comment by vasily-kirichenko** [comment]
It fails with
```
"E:\github\visualfsharp\src\fsharp\Fsc-proto\Fsc-proto.fsproj" (Build target) (4) ->
(CoreCompile target) ->
  E:\github\visualfsharp\src\absil\illib.fs(1106,51): error FS0039: The type 'ShardMap' is not defined in 'Intern al.Utilities.Collections'. [E:\github\visualfsharp\src\fsharp\Fsc-proto\Fsc-proto.fsproj]
  E:\github\visualfsharp\src\absil\illib.fs(1106,51): error FS0039: The type 'ShardMap' is not defined in 'Intern al.Utilities.Collections'. [E:\github\visualfsharp\src\fsharp\Fsc-proto\Fsc-proto.fsproj]
  E:\github\visualfsharp\src\absil\illib.fs(1106,6): error FS0035: This construct is deprecated: This type abbrev iation has one or more declared type parameters that do not appear in the type being abbreviated. Type abbreviati ons must use all declared type parameters in the type being abbreviated. Consider removing one or more type param eters, or use a concrete type definition that wraps an underlying type, such as 'type C<'a> = C of ...'. [E:\gith ub\visualfsharp\src\fsharp\Fsc-proto\Fsc-proto.fsproj]
  E:\github\visualfsharp\src\absil\illib.fs(1108,76): error FS0039: The type 'ShardMap' is not defined in 'Intern al.Utilities.Collections'. [E:\github\visualfsharp\src\fsharp\Fsc-proto\Fsc-proto.fsproj]

(a lot of similar errors)

```

**Comment by gerardtoconnor** [comment]
My merge before PR has messed something up apologies, I am looking into it now

File: src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs [hunk 1]
```diff
@@ -0,0 +1,1504 @@
...
+    new(count: int, kvps:('K * 'V) seq)  = 
+        let count,bucket = init<_,'K,'V>(count, kvps ,fun (k,v) -> KeyValuePair<'K,'V>(k,v)) in ShardMap<'K,'V>(count,bucket)
+    new(kvps:('K * 'V) seq)  = 
+        let mutable counter = 0
+        let mutable items = []
+        kvps |> Seq.iter (fun kvp -> 
+            counter <- counter + 1
+            items <- kvp :: items
+        )
+        ShardMap<'K,'V>(counter,kvps)
```
Comment (dsyme) [inline src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs]: items is unused after being initialized?

File: src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs [hunk 2]
```diff
@@ -0,0 +1,1504 @@
...
+open util
+open MapTree
+open System.Threading.Tasks
+
+    /// Shard Map
+    ////////////////////////////
+
+type ShardMap<'K,'V  when 'K : equality and 'K : comparison >(icount:int, nBucket:Shard<'K,'V> []) =
+    
+    let empty = newShardEmpty ()
```
Comment (dsyme) [inline src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs]: I don't see where this is used

File: src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs [hunk 3]
```diff
@@ -0,0 +1,1504 @@
...
+        let isize = bucket.Length
+        let nsize = isize * 2
+        resize nsize
+        
+            
+    let add(k:'K,v:'V) =
+        // let kh = k.GetHashCode()
+        // let bi = util.bucketIndex(kh,bucketBitMask)
+        // let si = util.shardIndex kh
+        let bi, si = getIndexes(k,bucketBitMask)
```
Comment (dsyme) [inline src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs]: Does the use of a tuple return here cause an unnecessary allocation?

File: src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs [hunk 4]
```diff
@@ -0,0 +1,1504 @@
...
+        (counter,newBucket)
+
+open util
+open MapTree
+open System.Threading.Tasks
+
+    /// Shard Map
+    ////////////////////////////
+
+type ShardMap<'K,'V  when 'K : equality and 'K : comparison >(icount:int, nBucket:Shard<'K,'V> []) =
```
Comment (dsyme) [inline src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs]: ``nBucket`` feels like an odd name for an array?

File: src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs [hunk 5]
```diff
@@ -0,0 +1,1504 @@
...
+    static member Collect (collectFn: 'T -> #seq<'V>) (keyFn:'V -> 'K) (collection:seq<'T>) =
+        
+        let initSize = Seq.length collection //:HACK
+        let size = if initSize < ShardSize then ShardSize else initSize
+        let comparer = LanguagePrimitives.FastGenericComparer<'K>
+        let bitdepth = calcBitMaskDepth size
+        let bucketSize = bucketSizeFromBitDepth (bitdepth)
+        let bucketBitMask = calcSubBitMask bitdepth
+        let newBucket = Array.zeroCreate<Shard<'K,'V>>(bucketSize)
+        let countRef = ref 0
```
Comment (dsyme) [inline src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs]: I believe this should be ``let mutable countRef`` and you should pass the address of ``countRef`` where appropriate?  That would save an allocation

File: src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs [hunk 6]
```diff
@@ -0,0 +1,1504 @@
...
+
+    /////////////////////////////////
+    /// Static Methods
+    /////////////////////////////////
+
+    static member inline Empty with get () = ShardMap<'K,'V>(0)
+
+    static member Collect (collectFn: 'T -> #seq<'V>) (keyFn:'V -> 'K) (collection:seq<'T>) =
+        
+        let initSize = Seq.length collection //:HACK
```
Comment (dsyme) [inline src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs]: this iterates the collection?

File: src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs (hunk 4 above)
Comment (gerardtoconnor) [reply src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs]: Apologies, naming conventions not set, my hash space array is made up of a jagged array so I called the internal  immutable slices "Shards", and the parent array holding the shards a Bucket (although the hash conflict MapTrees are more like traditional buckets). "nBucket" is just the temp naming of a "new Bucket" of shards to maintain immutability, happy to change this (and bucket naming convention for any actual merge).

File: src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs (hunk 2 above)
Comment (gerardtoconnor) [reply src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs]: There was an issue when I pulled & merged where some legacy artifact code was somehow reintroduced, apologies this is why I closed the PR, I realized the whole merge to the recent master had messed up a number of things.

File: src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs (hunk 3 above)
Comment (gerardtoconnor) [reply src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs]: I was starting a refactoring and would have preferred the single line allocation helper function, hoping the optimiser may lift this tupling out of the branch but I can revert back to commented out code if this is causing an unneeded allocation (or create a new struct type that has two int fields)

File: src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs (hunk 6 above)
Comment (gerardtoconnor) [reply src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs]: This was removed (and hack as per comment) but in PR due to merging issue apologies, please disregard

File: src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs (hunk 5 above)
Comment (gerardtoconnor) [reply src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs]: Thank you, noted, I will update in source

Comment (gerardtoconnor) [reply src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs]: image
updating means I am then unable to use `Interlocked.Increment` on the value so not sure if I can pass a ref to update if I use `let mutable countRef` ?

File: src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs (hunk 1 above)
Comment (gerardtoconnor) [reply src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs]: apologies, this still worked due to re-running seq iterator but the whole point was to run only once with count to pre-compute the hash space needed, I have updated in source as was an oversight.

**Comment by dsyme** [comment]
Do you think it would be possible to factor out the NameMap/StampMap part of this PR to "prepare the way" for experiments with improved collections?  

That is, we'd ideally like to have the codebase set up in a way where changing a collection implementation is less intrusive - either just changing an alias or a few lines somewhere.  Sometimes moving to more member/object programming helps with this

Also, when we identify the places where ordering is important we can note that via ``OrderedNameMap`` and so on.

**Comment by gerardtoconnor** [comment]
Thank you for the review @dsyme I apologize for the tidy up that is needed but hopefully, you can provide feedback on if you think this may have some application to the compiler where both lookup performance & immutability are needed? although it seems heavy allocating arrays, as big AVL trees require many path nodes to be added on single addition/removal, it turns out to be very performant without the need for rebalancing etc.

I will remove from NameMap/StampMap when I resubmit PR as requested, I appreciate what you mean on easy swapping in and out of collection types, I am also working on alternative `QueueList`.

If we can Identify when ordering (for fold/enum) is needed by `OrderedMap`, would be Ideal

Before you review any further please let me fix merge issue that has broken everything and put some code out of sync with my original repo at [MapTests](https://github.com/gerardtoconnor/MapTests/blob/master/)

**Comment by dsyme** [comment]
> Thank you for the review @dsyme I apologize for the tidy up that is needed but hopefully, you can provide feedback on if you think this may have some application to the compiler where both lookup performance & immutability are needed? although it seems heavy allocating arrays, as big AVL trees require many path nodes to be added on single addition/removal, it turns out to be very performant without the need for rebalancing etc.

I don't have a really strong grasp of the statistical properties of the collections dominating performance for the compiler,  I actually think that's the most important thing we need to understand first, e.g

* numbers of collections being created 
* the causes of those creations (i.e. causal stacks)
* their size
* the number of times they are are being accessed
* the number of linear accesses where the only operation is an "add"
* the key types for these collections

and so on.   Getting the compiler to the point that we can get full visibility on this would be awesome.

I'm sure we have very skewed distributions - some large collections being accessed often, and zillions of tiny collections.

**Pull request closed without merging**